File: vmlib/compiler/src/statements.c
#include <stdio.h>
#include "scanner.h"
#include "vm.h"
#include "compiler_defs.h"
#include "declarations.h"
#include "emitter.h"
#include "expressions.h"
#include "statements.h"
#include "import.h"
#include "variables.h"

void block(Parser *parser)
{
    match(parser, TOKEN_EOL);
    while (!check(parser, TOKEN_RIGHT_BRACE) && !check(parser, TOKEN_EOF))
    {
        declaration(parser);
    }

    match(parser, TOKEN_EOL); // optional end of line.
    consume(parser, TOKEN_RIGHT_BRACE, "Expect '}' after block.");
}

void expressionStatement(Parser *parser)
{
    expression(parser);
    if (!check(parser, TOKEN_EOF) && !check(parser, TOKEN_RIGHT_BRACE))
    {
        consume(parser, TOKEN_EOL, "Only one statement per line allowed");
    }
    emitByte(parser, OP_POP);
}

void forStatement(Parser *parser)
{
    beginScope(parser);
    consume(parser, TOKEN_LEFT_PAREN, "Expect '(' after 'for'.");
    if (check(parser, TOKEN_SEMI_COLON))
    {
        // No initializer.
    }
    else if (match(parser, TOKEN_VAR))
    {
        varDeclaration(parser);
    }
    else
    {
        expressionStatement(parser);
    }
    consume(parser, TOKEN_SEMI_COLON, "Expect ';' after loop intializer.");
    LoopCompiler loop = {0};
    loop.startAddress = getCurrentOffset(parser->currentFunction);
    loop.exitAddress = UNINITALISED_ADDRESS;
    loop.enclosing = parser->currentLoop;
    loop.loopScopeDepth = parser->currentFunction->scopeDepth;
    loop.breakJump = UNINITALISED_ADDRESS;
    parser->currentLoop = &loop;

    if (!match(parser, TOKEN_SEMI_COLON))
    {
        expression(parser);
        consume(parser, TOKEN_SEMI_COLON, "Expect ';' after loop condition.");

        // Jump out of the loop if the condition is false.
        loop.exitAddress = emitJump(parser, OP_JUMP_IF_FALSE);
        emitByte(parser, OP_POP); // Condition.
    }
    if (!match(parser, TOKEN_RIGHT_PAREN))
    {
        int bodyJump = emitJump(parser, OP_JUMP);

        int incrementStart = getCurrentOffset(parser->currentFunction);
        expression(parser);
        emitByte(parser, OP_POP);
        consume(parser, TOKEN_RIGHT_PAREN, "Expect ')' after for clauses.");

        emitLoop(parser);
        loop.startAddress = incrementStart;
        patchJump(parser, bodyJump);
    }

    statement(parser);

    emitLoop(parser);

    if (loop.breakJump != UNINITALISED_ADDRESS)
    {
        patchJump(parser, loop.breakJump);
    }
    if (loop.exitAddress != UNINITALISED_ADDRESS)
    {
        patchJump(parser, loop.exitAddress);
        emitByte(parser, OP_POP); // Condition.
    }

    endScope(parser);
    parser->currentLoop = parser->currentLoop->enclosing;
}

static void syntheticMethodCall(Parser *parser, const char *method_name)
{
    Token methodNameToken = syntheticToken(method_name);
    uint8_t constant = identifierConstant(parser, &methodNameToken);
    emitBytes(parser, OP_INVOKE, constant);
    emitByte(parser, 0); // zero arguments
}

void foreachStatement(Parser *parser)
{
    beginScope(parser);
    consume(parser, TOKEN_LEFT_PAREN, "Expect '(' after 'foreach'.");
    consume(parser, TOKEN_VAR, "Expect 'var' to declare foreach loop variable");
    Token loop_var_name = parser->current;
    uint8_t loop_var = parseVariable(parser, "Expect a variable name in the foreach loop");
    emitByte(parser, OP_NIL);
    defineVariable(parser, loop_var);

    consume(parser, TOKEN_IN, "Expect 'in' keyword in foreach loop");
    expression(parser);
    syntheticMethodCall(parser, "iterator");
    consume(parser, TOKEN_RIGHT_PAREN, "Expect ')' after 'foreach' condition.");

    Token iter_var_name = syntheticToken("");
    int iter_var = addLocal(parser, iter_var_name);
    emitBytes(parser, OP_SET_LOCAL, (uint8_t) iter_var);
    markInitialized(parser);

    LoopCompiler loop = {0};
    loop.startAddress = getCurrentOffset(parser->currentFunction);
    loop.exitAddress = UNINITALISED_ADDRESS;
    loop.breakJump = UNINITALISED_ADDRESS;
    loop.enclosing = parser->currentLoop;
    loop.loopScopeDepth = parser->currentFunction->scopeDepth;
    parser->currentLoop = &loop;

    emitBytes(parser, OP_GET_LOCAL, iter_var);
    syntheticMethodCall(parser, "has_next?");
    loop.exitAddress = emitJump(parser, OP_JUMP_IF_FALSE);
    emitByte(parser, OP_POP);

    emitBytes(parser, OP_GET_LOCAL, iter_var);
    syntheticMethodCall(parser, "get_next");
    int variable = resolveLocal(parser, parser->currentFunction, &loop_var_name);
    emitBytes(parser, OP_SET_LOCAL, (uint8_t) variable);
    emitByte(parser, OP_POP);

    statement(parser);

    emitLoop(parser);
    patchJump(parser, loop.exitAddress);
    if (loop.breakJump != UNINITALISED_ADDRESS)
    {
        patchJump(parser, loop.breakJump);
    }
    endScope(parser);
    emitByte(parser, OP_POP); // This feels weird, like I shouldn't need to do it.
    parser->currentLoop = parser->currentLoop->enclosing;
}

void ifStatement(Parser *parser)
{
    consume(parser, TOKEN_LEFT_PAREN, "Expect '(' after 'if'.");
    expression(parser);
    consume(parser, TOKEN_RIGHT_PAREN, "Expect ')' after condition.");

    int thenJump = emitJump(parser, OP_JUMP_IF_FALSE);
    emitByte(parser, OP_POP);
    statement(parser);

    int elseJump = emitJump(parser, OP_JUMP);
    patchJump(parser, thenJump);
    emitByte(parser, OP_POP);

    match(parser, TOKEN_EOL);
    if (match(parser, TOKEN_ELSE))
        statement(parser);
    patchJump(parser, elseJump);
}

void returnStatement(Parser *parser)
{
    if (parser->currentFunction->type == TYPE_SCRIPT)
    {
        error(parser, "Cannot return from top-level code.");
    }
    if (match(parser, TOKEN_EOL))
    {
        emitReturn(parser);
    }
    else
    {
        if (parser->currentFunction->type == TYPE_INITIALIZER)
        {
            error(parser, "Cannot return a value from an initializer.");
        }
        expression(parser);
        if (!check(parser, TOKEN_RIGHT_BRACE))
        {
            consume(parser, TOKEN_EOL, "Only one statement per line allowed");
        }
        emitByte(parser, OP_RETURN);
    }
}

void whileStatement(Parser *parser)
{
    LoopCompiler loop = {0};
    loop.startAddress = getCurrentOffset(parser->currentFunction);
    loop.exitAddress = UNINITALISED_ADDRESS;
    loop.enclosing = parser->currentLoop;
    loop.breakJump = UNINITALISED_ADDRESS;
    loop.loopScopeDepth = parser->currentFunction->scopeDepth;
    parser->currentLoop = &loop;
    consume(parser, TOKEN_LEFT_PAREN, "Expect '(' after 'while'.");
    expression(parser);
    consume(parser, TOKEN_RIGHT_PAREN, "Expect ')' after condition.");

    loop.exitAddress = emitJump(parser, OP_JUMP_IF_FALSE);

    emitByte(parser, OP_POP);
    statement(parser);

    emitLoop(parser);

    patchJump(parser, loop.exitAddress);
    if (loop.breakJump != UNINITALISED_ADDRESS)
    {
        patchJump(parser, loop.breakJump);
    }
    emitByte(parser, OP_POP);
    parser->currentLoop = parser->currentLoop->enclosing;
}

static void patchAddress(Parser *parser, int offset)
{
    int currentOffset = getCurrentOffset(parser->currentFunction);
    setCodeOffset(parser->currentFunction, offset, (currentOffset >> 8) & 0xff);
    setCodeOffset(parser->currentFunction, offset + 1, currentOffset & 0xff);
}

void tryStatement(Parser *parser)
{
    emitByte(parser, OP_PUSH_EXCEPTION_HANDLER);
    int exceptionType = getCurrentOffset(parser->currentFunction);
    emitByte(parser, 0xff);
    int handlerAddress = getCurrentOffset(parser->currentFunction);
    emitBytes(parser, 0xff, 0xff);
    int finallyAddress = getCurrentOffset(parser->currentFunction);
    emitBytes(parser, 0xff, 0xff);

    statement(parser);

    emitByte(parser, OP_POP_EXCEPTION_HANDLER);
    int successJump = emitJump(parser, OP_JUMP);
    match(parser, TOKEN_EOL);

    if (match(parser, TOKEN_CATCH))
    {
        beginScope(parser);
        consume(parser, TOKEN_LEFT_PAREN, "Expect '(' after catch");
        consume(parser, TOKEN_IDENTIFIER, "Expect type name to catch");
        uint8_t name = identifierConstant(parser, &parser->previous);
        setCodeOffset(parser->currentFunction, exceptionType, name);
        patchAddress(parser, handlerAddress);
        if (match(parser, TOKEN_AS))
        {
            consume(parser, TOKEN_IDENTIFIER, "Expect identifier for exception instance");
            uint8_t ex_var = addLocal(parser, parser->previous);
            markInitialized(parser);
            emitBytes(parser, OP_SET_LOCAL, ex_var);
        }
        consume(parser, TOKEN_RIGHT_PAREN, "Expect ')' after catch statement");
        emitByte(parser, OP_POP_EXCEPTION_HANDLER);
        statement(parser);
        match(parser, TOKEN_EOL);
        endScope(parser);
    }
    patchJump(parser, successJump);

    if (match(parser, TOKEN_FINALLY))
    {
        // If we arrive here from either the try or handler blocks, then we don't
        // want to continue propagating the exception
        emitByte(parser, OP_FALSE);

        patchAddress(parser, finallyAddress);
        statement(parser);

        int continueExecution = emitJump(parser, OP_JUMP_IF_FALSE);
        emitByte(parser, OP_POP); // Pop the bool off the stack
        emitByte(parser, OP_PROPAGATE_EXCEPTION);
        patchJump(parser, continueExecution);
        emitByte(parser, OP_POP);
    }
}

void throwStatement(Parser *parser)
{
    expression(parser);
    if (!check(parser, TOKEN_EOF))
        consume(parser, TOKEN_EOL, "Only one statement per line allowed");
    emitByte(parser, OP_THROW);
}

void rethrowStatement(Parser *parser)
{
    expression(parser);
    if (!check(parser, TOKEN_EOF))
        consume(parser, TOKEN_EOL, "Only one statement per line allowed");
    emitByte(parser, OP_PROPAGATE_EXCEPTION);
}

void importStatement(Parser *parser)
{
    expression(parser);
    consume(parser, TOKEN_AS, "Expected 'as' after the module to import");
    emitByte(parser, OP_IMPORT);
    // Imports are a function that return NIL, so ditch the nil from the stack
    emitByte(parser, OP_POP);
    uint8_t global = parseVariable(parser, "Expected a variable name for the imported module.");
    defineVariable(parser, global);
}

void nextStatement(Parser *parser)
{
    if (parser->currentLoop == NULL) {
        error(parser, "Can't use 'next' outside of a loop.");
        return;
    }

    // Discard any locals created inside the loop.
    for (int i = parser->currentFunction->localCount - 1;
        i >= 0 && parser->currentFunction->locals[i].depth > parser->currentLoop->loopScopeDepth;
        i--) {
        emitByte(parser, OP_POP);
    }

    // Jump to top of current innermost loop.
    emitLoop(parser);
}

static void breakStatement(Parser *parser)
{
    if (parser->currentLoop == NULL) {
        error(parser, "Can't use 'break' outside of a loop.");
        return;
    }

    if (parser->currentLoop->breakJump != UNINITALISED_ADDRESS)
    {
        error(parser, "Only one break statement per loop is supported.");
        return;
    }

    parser->currentLoop->breakJump = emitJump(parser, OP_JUMP);
}

void statement(Parser *parser)
{
    match(parser, TOKEN_EOL);
    if (match(parser, TOKEN_FOR))
    {
        forStatement(parser);
    }
    else if (match(parser, TOKEN_FOREACH))
    {
        foreachStatement(parser);
    }
    else if (match(parser, TOKEN_IF))
    {
        ifStatement(parser);
    }
    else if (match(parser, TOKEN_RETURN))
    {
        returnStatement(parser);
    }
    else if (match(parser, TOKEN_WHILE))
    {
        whileStatement(parser);
    }
    else if (match(parser, TOKEN_TRY))
    {
        tryStatement(parser);
    }
    else if (match(parser, TOKEN_RETHROW))
    {
        rethrowStatement(parser);
    }
    else if (match(parser, TOKEN_THROW))
    {
        throwStatement(parser);
    }
    else if (match(parser, TOKEN_LEFT_BRACE))
    {
        beginScope(parser);
        block(parser);
        endScope(parser);
    }
    else if (match(parser, TOKEN_IMPORT))
    {
        importStatement(parser);
    }
    else if (match(parser, TOKEN_NEXT))
    {
        nextStatement(parser);
    }
    else if (match(parser, TOKEN_BREAK)) {
        breakStatement(parser);
    }
    else
    {
        expressionStatement(parser);
    }
}

File: sharp-comet/compiler/statements.cs
using sharpcomet.lexer;
using sharpcomet.vmlib;
using System;

namespace sharpcomet.compiler;

public partial class Parser
{
    private void EmitLoop()
    {
        if (!CurrentFunction.EmitLoop(CurrentLoop!.StartAddress))
        {
            Error("Loop body too large.");
        }
    }

    private void PatchJump(int jump)
    {
        if (!CurrentFunction.PatchJump(jump))
        {
            Error("Too much code to jump over!");
        }
    }

    private void ForStatement()
    {
        CurrentFunction.BeginScope();
        Consume(TokenType.LeftParen, "Expected '(' after 'for'.");
        if (Check(TokenType.SemiColon))
        {
            // No initializer.
        }
        else if (Match(TokenType.Var))
        {
            VarDeclaration();
        }
        else
        {
            ExpressionStatement();
        }
        Consume(TokenType.SemiColon, "Expected ';' after loop intializer.");
        CurrentLoop = new LoopCompiler(CurrentFunction.CurrentOffset, CurrentFunction.ScopeDepth, CurrentLoop);

        if (!Match(TokenType.SemiColon))
        {
            Expression();
            Consume(TokenType.SemiColon, "Expected ';' after loop condition.");

            // Jump out of the loop if the condition is false.
            CurrentLoop.ExitAddress = CurrentFunction.EmitJump(Op.JumpIfFalse);
            CurrentFunction.EmitBytes(Op.Pop); // Condition.
        }
        if (!Match(TokenType.RightParen))
        {
            int bodyJump = CurrentFunction.EmitJump(Op.Jump);

            int incrementStart = CurrentFunction.CurrentOffset;
            Expression();
            CurrentFunction.EmitBytes(Op.Pop);
            Consume(TokenType.RightParen, "Expected ')' after for clauses.");

            EmitLoop();

            CurrentLoop.StartAddress = incrementStart;
            PatchJump(bodyJump);
        }

        Statement();

        EmitLoop();

        if (CurrentLoop.BreakJump != null)
        {

            PatchJump(CurrentLoop.BreakJump.Value);
        }
        if (CurrentLoop.ExitAddress != null)
        {
            PatchJump(CurrentLoop.ExitAddress.Value);
            CurrentFunction.EmitBytes(Op.Pop);
        }

        CurrentFunction.EndScope();
        CurrentLoop = CurrentLoop.Enclosing;
    }

    private void IfStatement()
    {
        Consume(TokenType.LeftParen, "Expected '(' after 'if'.");
        Expression();
        Consume(TokenType.RightParen, "Expected ')' after condition.");

        int thenJump = CurrentFunction.EmitJump(Op.JumpIfFalse);
        CurrentFunction.EmitBytes(Op.Pop);
        Statement();

        int elseJump = CurrentFunction.EmitJump(Op.Jump);
        PatchJump(thenJump);
        CurrentFunction.EmitBytes(Op.Pop);

        Match(TokenType.EndOfLine);
        if (Match(TokenType.Else))
            Statement();

        PatchJump(elseJump);
    }

    private void Block()
    {
        Match(TokenType.EndOfLine);
        while (!Check(TokenType.RightBrace) && !Check(TokenType.EndOfFile))
        {
            Declaration();
        }
        Match(TokenType.EndOfLine);
        Consume(TokenType.RightBrace, "Expected a '}' after a block.");
    }

    private void ExpressionStatement()
    {
        Expression();
        if (!Check(TokenType.EndOfFile) && !Check(TokenType.RightBrace))
        {
            Consume(TokenType.EndOfLine, "Only one statement per line allowed.");
        }
        CurrentFunction.EmitBytes(Op.Pop);
    }

    private void ThrowStatement()
    {
        Expression();
        if (!Check(TokenType.EndOfFile))
            Consume(TokenType.EndOfLine, "Only one statement per line allowed.");
        CurrentFunction.EmitBytes(Op.Throw);
    }

    private void RethrowStatement()
    {
        Expression();
        if (!Check(TokenType.EndOfFile))
            Consume(TokenType.EndOfLine, "Only one statement per line allowed.");
        CurrentFunction.EmitBytes(Op.Rethrow);
    }

    private void SyntheticMethodCall(string methodName)
    {
        Token methodNameToken = SyntheticToken(methodName);
        byte constant = IdentifierConstant(methodNameToken);
        CurrentFunction.EmitBytes((byte)Op.Invoke, constant, 0); // zero arguments
    }

    private void ForEachStatement()
    {
        CurrentFunction.BeginScope();
        Consume(TokenType.LeftParen, "Expected '(' after 'foreach'.");
        Consume(TokenType.Var, "Expected 'var' to declare foreach loop variable.");
        Token loopVarName = Current;
        byte loopVar = ParseVariable("Expected a variable name in the foreach loop.");
        CurrentFunction.EmitBytes(Op.Nil);
        CurrentFunction.DefineVariable(loopVar);

        Consume(TokenType.In, "Expected 'in' keyword in foreach loop.");
        Expression();
        SyntheticMethodCall("iterator");
        Consume(TokenType.RightParen, "Expected ')' after 'foreach' condition.");

        byte iterVar = CurrentFunction.AddLocal(string.Empty);
        CurrentFunction.EmitBytes((byte)Op.SetLocal, iterVar);
        CurrentFunction.MarkInitialized();

        CurrentLoop = new LoopCompiler(CurrentFunction.CurrentOffset, CurrentFunction.ScopeDepth, CurrentLoop);

        CurrentFunction.EmitBytes((byte)Op.GetLocal, iterVar);
        SyntheticMethodCall("has_next?");
        CurrentLoop.ExitAddress = CurrentFunction.EmitJump(Op.JumpIfFalse);
        CurrentFunction.EmitBytes(Op.Pop);

        CurrentFunction.EmitBytes((byte)Op.GetLocal, iterVar);
        SyntheticMethodCall("get_next");
        int variable = CurrentFunction.ResolveLocal(loopVarName.Representation);
        CurrentFunction.EmitBytes((byte)Op.SetLocal, (byte)variable, (byte)Op.Pop);

        Statement();

        EmitLoop();
        PatchJump(CurrentLoop.ExitAddress.Value);
        if (CurrentLoop.BreakJump != null)
        {
            PatchJump(CurrentLoop.BreakJump.Value);
        }
        CurrentFunction.EndScope();
        CurrentFunction.EmitBytes(Op.Pop); // This feels weird, like I shouldn't need to do it.
        CurrentLoop = CurrentLoop.Enclosing;
    }

    private void ReturnStatement()
    {
        if (CurrentFunction.IsScript())
        {
            Error("Cannot return from top-level code.");
        }
        if (Match(TokenType.EndOfLine))
        {
            CurrentFunction.EmitReturn();
        }
        else
        {
            if (CurrentFunction.IsInitializer())
            {
                Error("Cannot return from an initializer.");
            }
            Expression();
            if (!Check(TokenType.RightBrace))
            {
                Consume(TokenType.EndOfLine, "Only one statement per line allowed.");
            }
            CurrentFunction.EmitBytes(Op.Return);
        }
    }

    private void WhileStatement()
    {
        CurrentLoop = new LoopCompiler(CurrentFunction.CurrentOffset, CurrentFunction.ScopeDepth, CurrentLoop);
        Consume(TokenType.LeftParen, "Expected '(' after 'while'.");
        Expression();
        Consume(TokenType.RightParen, "Expected ')' after condition.");

        CurrentLoop.ExitAddress = CurrentFunction.EmitJump(Op.JumpIfFalse);

        CurrentFunction.EmitBytes(Op.Pop);
        Statement();
        EmitLoop();

        PatchJump(CurrentLoop.ExitAddress.Value);
        if (CurrentLoop.BreakJump.HasValue)
        {
            PatchJump(CurrentLoop.BreakJump.Value);
        }
        CurrentFunction.EmitBytes(Op.Pop);
        CurrentLoop = CurrentLoop.Enclosing;
    }

    private void TryStatement()
    {
        CurrentFunction.EmitBytes(Op.PushExceptionHandler);
        int exceptionType = CurrentFunction.CurrentOffset;
        CurrentFunction.EmitBytes(0xFF);
        int handlerAddress = CurrentFunction.CurrentOffset;
        CurrentFunction.EmitBytes(0XFF, 0XFF);
        int finallyAddress = CurrentFunction.CurrentOffset;
        CurrentFunction.EmitBytes(0XFF, 0XFF);

        Statement();

        CurrentFunction.EmitBytes(Op.PopExceptionHandler);
        int successJump = CurrentFunction.EmitJump(Op.Jump);
        Match(TokenType.EndOfLine);
        bool tryBlockCompleted = false;

        if (Match(TokenType.Catch))
        {
            tryBlockCompleted = true;
            CurrentFunction.BeginScope();
            Consume(TokenType.LeftParen, "Expected '(' after catch.");
            Consume(TokenType.Identifier, "Expected a type name to catch.");
            byte name = IdentifierConstant(Previous);
            CurrentFunction.SetCodeOffset(exceptionType, name);
            CurrentFunction.PatchAddress(handlerAddress);
            if (Match(TokenType.As))
            {
                Consume(TokenType.Identifier, "Expected an identifier for the exception instance.");
                byte exVar = CurrentFunction.AddLocal(Previous.Representation);
                CurrentFunction.MarkInitialized();
                CurrentFunction.EmitBytes((byte)Op.SetLocal, exVar);
            }
            Consume(TokenType.RightParen, "Expected ')' after a catch statement.");
            CurrentFunction.EmitBytes(Op.PopExceptionHandler);
            Statement();
            Match(TokenType.EndOfLine);
            CurrentFunction.EndScope();
        }
        PatchJump(successJump);

        if (Match(TokenType.Finally))
        {
            tryBlockCompleted = true;
            // If we arrive here from either the try or handler blocks, then we don't
            // want to continue propagating the exception
            CurrentFunction.EmitBytes(Op.False);

            CurrentFunction.PatchAddress(finallyAddress);
            Statement();

            int continueExecution = CurrentFunction.EmitJump(Op.JumpIfFalse);
            CurrentFunction.EmitBytes(Op.Pop); // Pop the bool off the stack
            CurrentFunction.EmitBytes(Op.PropagateException);
            PatchJump(continueExecution);
            CurrentFunction.EmitBytes(Op.Pop);
        }

        if (tryBlockCompleted == false)
        {
            ErrorAtCurrent("A try statement requires a catch, finally or both");
        }
    }

    private void ImportStatement()
    {
        byte? importParamCount = 0;
        if (Match(TokenType.LeftBrace))
        {
            byte[] moduleIdentifierConstants = new byte[255];
            if (Match(TokenType.Star))
            {
                importParamCount = null;
            }
            else
            {
                do
                {
                    if (importParamCount == 255)
                    {
                        ErrorAtCurrent("Cannot have more than 255 import parameters.");
                    }
                    Match(TokenType.EndOfLine);
                    Consume(TokenType.Identifier, "Expected a module import identifier.");
                    byte paramConstant = IdentifierConstant(Previous);
                    moduleIdentifierConstants[importParamCount.Value] = paramConstant;
                    importParamCount++;

                } while (Match(TokenType.Comma));
            }
            Consume(TokenType.RightBrace, "Expected a '}' after import parameters.");
            Match(TokenType.EndOfLine);
            Consume(TokenType.From, "Expected 'from' after import parameters.");
            Expression();
            // Imports are a function that returns nil, so pop that off the stack, too
            CurrentFunction.EmitBytes(Op.Import, Op.Pop);
            CurrentFunction.EmitBytes((byte)Op.ImportParams, importParamCount ?? 0);
            if (importParamCount != null)
            {
                CurrentFunction.EmitBytes(moduleIdentifierConstants);
            }
        }
        else
        {
            Expression();
            // Imports are a function that returns nil, so pop that off the stack, too
            CurrentFunction.EmitBytes(Op.Import, Op.Pop);
            Consume(TokenType.As, "Expected 'as' after the module to import.");
            byte global = ParseVariable("Expected a variable name for the imported module.");
            CurrentFunction.DefineVariable(global);
        }
    }

    private void NextStatement()
    {
        if (CurrentLoop == null) {
            Error("Can't use 'next' outside of a loop.");
            return;
        }

        // Discard any locals created inside the loop.
        CurrentFunction.DiscardCurrentScope(CurrentLoop.LoopScopeDepth);

        // Jump to top of current innermost loop.
        EmitLoop();
    }

    private void BreakStatement()
    {
        if (CurrentLoop == null)
        {
            Error("Can't use 'break' outside of a loop.");
            return;
        }

        if (CurrentLoop.BreakJump != null)
        {
            Error("Only one break statement per loop is supported.");
            return;
        }

        // Discard any locals created inside the loop.
        CurrentFunction.DiscardCurrentScope(CurrentLoop.LoopScopeDepth);

        CurrentLoop.BreakJump = CurrentFunction.EmitJump(Op.Jump);
    }

    private void Statement()
    {
        Match(TokenType.EndOfLine);
        if (Match(TokenType.For))
        {
            ForStatement();
        }
        else if (Match(TokenType.ForEach))
        {
            ForEachStatement();
        }
        else if (Match(TokenType.If))
        {
            IfStatement();
        }
        else if (Match(TokenType.Return))
        {
            ReturnStatement();
        }
        else if (Match(TokenType.While))
        {
            WhileStatement();
        }
        else if (Match(TokenType.Try))
        {
            TryStatement();
        }
        else if (Match(TokenType.Rethrow))
        {
            RethrowStatement();
        }
        else if (Match(TokenType.Throw))
        {
            ThrowStatement();
        }
        else if (Match(TokenType.LeftBrace))
        {
            CurrentFunction.BeginScope();
            Block();
            CurrentFunction.EndScope();
        }
        else if (Match(TokenType.Import))
        {
            ImportStatement();
        }
        else if (Match(TokenType.Next))
        {
            NextStatement();
        }
        else if (Match(TokenType.Break))
        {
            BreakStatement();
        }
        else
        {
            ExpressionStatement();
        }
    }
}